### ISSUE TITLE
[2018 CVPR] Jointly Optimize Data Augmentation and Network Training: Adversarial Data Augmentation in Human Pose Estimation

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#3]
Human pose estimation에서 사용되는 heuristic random data augmentation은 3가지 문제가 있음

1. 모든 training set에 같은 random data augmentation을 적용하여 너무 어렵거나 너무 쉬운 ineffective variations로 인해 네트워크 학습에 영향을 줌 (ineffective variations that are either too 'hard' or too 'easy')
2. Random data augmentation은 static distribution에서 sample하기에 dynamic training status와 맞지 않음
3. 흔히 사용되는 Gaussian distribution은 rare하지만 useful augmentation을 sample할 수 있는 기회가 적기에, long-tail issue를 해결할 수 없음

이에 저자들은 random data augmentation이 아닌 학습 가능한 data augmentation과 더불어, augmentation과 network를 따로 학습하는 것이 아닌 jointly optimize하는 방법을 제시함

해당 방법은 human pose estimation뿐만 아니라 face alignment, instance segmentation과 같은 vision task에서도 사용 가능하다고 함

GAN과 같은 방식으로 학습을 진행함
![Network](https://user-images.githubusercontent.com/49643709/110061354-d1a38900-7daa-11eb-84a4-7daf600b71d4.png)
![Generator Loss](https://user-images.githubusercontent.com/49643709/110059118-f564d000-7da6-11eb-9bdd-df972681d3ef.png)
![Discriminator Loss](https://user-images.githubusercontent.com/49643709/110059077-debe7900-7da6-11eb-8d27-abb051185424.png)

- Generator인 augmentation network가 augmentation operation에 대한 adversarial distribution 생성
- Adversarial distribution에서 sample한 값을 이용하여 adversarial augmented image 생성
- Discriminator인 pose network가 random augmented image와 adversarial augmented image의 loss를 각각 계산
- Adversarial augmented image loss가 random augmented image loss보다 커지도록 generator 학습
- Discriminator는 adversarial augmented image를 학습

Discriminator는 U-net design의 architecture을 이용하며, generator는 discriminator의 features가 input인 encoder architecture을 사용한다 (bridged features of U-net as inputs instead of raw images)

![Augmentation Network](https://user-images.githubusercontent.com/49643709/110061425-eb44d080-7daa-11eb-8b40-67e563f52bfd.png)

3개의 augmentation만 사용함 (scaling, rotating, occlusion)

Rotation과 scaling은 ground truth의 transformation에 영향을 주기에 raw image에 적용하며, occlusion은 ground truth의 transformation에 영향을 주지 않기에 deep features에 적용함 (it is more effective to occlude deep features instead of image pixels)

Augmentation을 적용하는 대상이 다르기에 ASR (adversarial scaling and rotating), AHO (adversarial hierarchical occluding) 2가지로 나눔

ASR (scale, rotation)
- Augmentation을 직접 생성하면 대부분의 경우 upside-down augmentation만 학습 (direct generation would collapse the training)
- 이를 방지하기 위해 n bin histogram 생성을 학습 (n_scaling = 7, n_rotating=9)
- Augmentation network가 histogram에서 bin을 선택
- 선택한 bin을 mean으로 하는 Gaussian distribution에서 sample
- Sample value를 통해 adversarial augmented image 생성
- ASR는 pre-train한 후 pose network와 joint training함 (crucial to pre-train the augmentation network)
  ![KL-divergence loss](https://user-images.githubusercontent.com/49643709/110074285-4a630f00-7dc4-11eb-9e41-9f08c4d4acf2.png)
  - Ground truth histogram과 KL-divergence loss를 통해 학습
- Ground truth histogram은 다음과 같은 방법을 통해 구함
  1. 훈련 이미지에 대하여 n_scaling * n_rotating augmented images 생성 (각 bin에서 sample한 값들의 combination)
  2. Pose network를 이용해 augmented images loss를 normalize한 후 축적하여 ground truth histogram 생성

AHO (occlusion)
- Deep features occlusion이 image pixel occlusion보다 효과적임
- Lowest resolution feature에 4 * 4 mask occlude하며 feature resolution이 높아질수록 같은 자리에 mask를 scale-up하여 occlude
- ASR과 마찬가지로 instance occluding mask를 생성하지 않고 occluding distribution을 생성
- AHO는 pre-train한 후 pose network와 joint training함 (crucial to pre-train the augmentation network)
  ![KL-divergence](https://user-images.githubusercontent.com/49643709/110082592-e5fa7c80-7dd0-11eb-8803-a8d3650f7a37.png)
  - Ground truth occluding distribution과 KL-divergence loss를 통해 학습
- Ground truth occluding distribution은 다음과 같은 방법을 통해 구함
  1. Image를 w * h grid로 나눔 (w=h=4)
  2. 해당 grid에 속해있는 joint의 개수를 count
  3. Grid를 normalize하여 ground truth occluding distribution 생성

Generator는 reward-penalty policy를 이용하여 학습하며, direct operation 대신 distribution을 생성하는 것이 outlier에 더 robust하다고 함
- Adversarial augmentation이 random augmentation보다 loss가 크다면 sample된 bin (ASR), cell (AHO) 확률 증가 (reward)
- Adversarial augmentation이 random augmentation보다 loss가 작다면 sample된 bin (ASR), cell (AHO) 확률 감소 (penalize)
![image](https://user-images.githubusercontent.com/49643709/110083146-b0a25e80-7dd1-11eb-93cd-98201ee03886.png)

Joint training
- ASR, ASO를 동시에 적용하면 너무 어려워, pose network가 학습을 못함
- Mini-batch X를 동등하게 3등분하여 하나는 random data augmentation, 하나는 ASR, 하나는 AHO를 적용하여 학습을 진행함
  ![Train](https://user-images.githubusercontent.com/49643709/110091216-8fdf0680-7ddb-11eb-8059-68dadfbb7789.png)
  ![Train - Adversarial Augmented](https://user-images.githubusercontent.com/49643709/110091231-966d7e00-7ddb-11eb-820c-c6220004fd4b.png)

결과 및 참고할만한 figure는 다음과 같음
![image](https://user-images.githubusercontent.com/49643709/110096776-a4260200-7de1-11eb-812d-9335c9741077.png)
![image](https://user-images.githubusercontent.com/49643709/110096863-bc961c80-7de1-11eb-9667-6e49d90b0d0c.png)
![image](https://user-images.githubusercontent.com/49643709/110096830-b142f100-7de1-11eb-979f-92ee58d70cdb.png)
![image](https://user-images.githubusercontent.com/49643709/110096919-c9b30b80-7de1-11eb-96ea-6bec1530bc3c.png)
![image](https://user-images.githubusercontent.com/49643709/110096926-ccadfc00-7de1-11eb-9507-f38446e115f4.png)

**Comment by Jasonlee1995** [comment]
간단하게 정리 + 빠진 부분 채우면 다음과 같음

## Network
### 1. Generator : Augmentation Network (scaling, rotating, occluding)
#### 1.1. Adversarial Scaling and Rotating (ASR) 
- Pose network를 통해 image의 feature 추출
- Feature를 통해 scale, rotation distribution 생성 (histogram이므로 encoder + linear layer로 bin에 대한 확률값 얻어냄)
- Scale, rotation distribution에서 bin을 각각 sample
- Sample한 bin을 mean으로 하는 Gaussian distribution을 생성하여 value sample
- 2개의 value를 이용하여 image scale, rotate augment 적용
#### 1.2. Adversarial Hierarchical Occluding (AHO) 
- Pose network를 통해 image의 feature 추출
- Feature를 통해 heatmap 생성
- Heatmap을 통해 feature에 1 ~ 2개의 4 * 4 mask 생성
### 2. Discriminator : Pose Network (U-net design architecture)

## Train
- 먼저 random augmented image를 통해 discriminator 학습 (few epochs + lr 0.00025)
  * crop each person + random scaling (0.75~1.25) + random rotating (-/+ 30 deg) + random flipping + resize 256 * 256
- Pre-train generator (ASR, AHQ) with freezing discriminator (KL-divergence loss + lr 0.00025)
- Jointly train generator + discriminator (mini-batch 3등분하여 random augment, ASR, AHQ + lr 0.00005 + Reduce lr on Plateau + reward and penalty policy on generator)
- RMSProp optimizer
